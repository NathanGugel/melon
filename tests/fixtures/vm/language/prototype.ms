let Animal = {
    name = null,
    create = |name| => {
        return { name = name };
    },
    walk = -> {
        let walkingStr = "Walking";
        io.print(walkingStr .. " " .. this.name);
    },
    introduce = func intro |otherName| -> {
        io.print("Hey " .. otherName .. ", this is " .. this.name .. "!");
    }
};

let Dog = {
    create = |name| => {
        return Animal.create(name) @ Dog;
    },

    bark = -> {
        io.print("Woof!!");
    }
} @ Animal;

let Shiba = {
    create = |name| => {
        return Dog.create(name) @ Shiba;
    },

    bark = -> {
        io.print("Woof with pride!!");
    }
} @ Dog;

func dogRoutine|dog| => {
    dog->walk();
    dog->introduce("Tester");
    dog->bark();
    dog.bark(dog);
}

let myDog = Dog.create("Fido");
let myShiba = Shiba.create("Shishi");

dogRoutine(myDog);
dogRoutine(myShiba);

io.print(@myDog == Dog);
io.print(@myShiba == Dog);
io.print(@myShiba == Shiba);
io.print(@@myShiba == Dog);

let unblessedObject = {};
io.print(@unblessedObject);
io.print(@{});
