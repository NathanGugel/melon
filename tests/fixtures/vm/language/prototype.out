Walking Fido
Hey Tester, this is Fido!
Woof!!
Woof!!
Walking Shishi
Hey Tester, this is Shishi!
Woof with pride!!
Woof with pride!!
true
false
true
true
null
null
